src/TabComplete.js

```diff
@@ -0,0 +1,306 @@
+        this.inPassiveMode = ev.passive;
+
+        if (ev.keyCode !== KEY_TAB) {
+            if (this.completing && ev.keyCode !== KEY_SHIFT &&
+                    !ev.metaKey && !ev.ctrlKey && !ev.altKey && ev.keyCode !== KEY_WINDOWS) {
+                // they're resuming typing; reset tab complete state vars.
+                this.stopTabCompleting();
+
+                // fall through to auto-enter-tab-completing if set
+                if (!this.opts.autoEnterTabComplete) {
+                    return true;
+                }
+            }
+
+            if (this.opts.autoEnterTabComplete) {
+                clearTimeout(this.enterTabCompleteTimerId);
+                this.enterTabCompleteTimerId = setTimeout(() => {
+                    if (!this.completing) {
+                        // inject a fake tab event so we use the same code paths
+                        // as much as possible. Add a 'passive' flag to indicate
+                        // that they didn't really hit this key.
+                        this.onKeyDown({
+                            keyCode: KEY_TAB,
+                            passive: true,
+                            preventDefault: function(){} // NOP
+                        })
+                    }
+                }, DELAY_TIME_MS);
+            }
+
+            return false;
+        }
+
+        if (!this.textArea) {
+            console.error("onKeyDown called before a <textarea> was set!");
+            return false;
+        }
+
+        // init struct if necessary
+        if (!this.completing) {
+            this.startTabCompleting();
+        }
+
+        if (ev.shiftKey) {
+            this.nextMatchedEntry(-1);
+        }
+        else {
+            // if we were in passive mode we got out of sync by incrementing the
+            // index to show the peek view but not set the text area. Therefore,
+            // we want to set the *current* index rather than the *next* index.
+            this.nextMatchedEntry(wasInPassiveMode ? 0 : 1);
+        }
+
+        // prevent the default TAB operation (typically focus shifting)
+        ev.preventDefault();
+        this._notifyStateChange();
+        return true;
+    }
+
+    /**
+     * Set the textarea to the next value in the matched list.
+     * @param {Number} offset Offset to apply *before* setting the next value.
+     */
+    nextMatchedEntry(offset) {
+        if (this.matchedList.length === 0) {
+            return;
+        }
+
+        // work out the new index, wrapping if necessary.
+        this.currentIndex += offset;
+        if (this.currentIndex >= this.matchedList.length) {
+            this.currentIndex = 0;
+        }
+        else if (this.currentIndex < 0) {
+            this.currentIndex = this.matchedList.length - 1;
+        }
+        var looped = (
+            // impossible to loop if they've never hit tab; catch forward and backward looping
+            !this.inPassiveMode && this.currentIndex === 0
```

Comment:
> isn't this true if I return to the first entry with shift-tab?

The first entry is the partial word that is being tab-completed and not the first suggestion.

> Why is this better than setting looped=true in the two if clauses at :213 and :216?

Because that is subtly wrong. Setting `looped=true` on the `if (this.currentIndex < 0)` clause would be saying you've looped when transitioning to the last entry via a shift-tab at the start of a tab completion. In this case we do *not* want to cancel the tab complete nor flash red to say no more entries. It's more accurate to state that `looped` is "going to display the original partial text" to avoid confusion over which transition points count as a loop (0 -> n-1 vs 0 -> 1), so will tweak wording on this.